### PR TITLE
Adds a `maximize` flag to Adam

### PR DESCRIPTION
**Pull request by OliverFM**
Solves the next most important use case in #68052.

I have kept the style as close to that in SGD as seemed reasonable, given the slight differences in their internal implementations.

All feedback welcome!

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### CODE REVIEW

**Comment by OliverFM** [comment]
@albanD I wanted to follow up on this one again -- I have merged in the nightly build and everything passes locally. I am not sure if the ROCM tests are still going to causes issues though.

Anything else I should do with this PR?

**Comment by albanD** [comment]
I actually have the same question as https://github.com/pytorch/pytorch/pull/68579#discussion_r752217573 for this one.
Why is flipping the lr a better choice than flipping the gradients here?

**Comment by OliverFM** [comment]
> I actually have the same question as [#68579 (comment)](https://github.com/pytorch/pytorch/pull/68579#discussion_r752217573) for this one. Why is flipping the lr a better choice than flipping the gradients here?

Good question.
I prefer it because it confines the maximise logic to the end -- also, checking [the docs](https://pytorch.org/docs/stable/generated/torch.optim.Adam.html#torch.optim.Adam) I think that they should be equivalent. Since `v_t` would be unaffected, and `m_t` would be multiplied by `-1` and that change would carry on to the end.

I think that the outcome of flipping the gradients would be the same, but flipping the LR seems clearer to me.

I also see now that I need to update the docs in the PR too.

**Comment by OliverFM** [comment]
@albanD does this seem reasonable?

**Comment by albanD** [comment]
Sorry I was on holidays and it took some time to answer.

> also, checking the docs I think that they should be equivalent.

I am not sure. The weight decay in particular looks like it won't be applied properly?

**Comment by OliverFM** [comment]
> Sorry I was on holidays and it took some time to answer.
> 
> > also, checking the docs I think that they should be equivalent.
> 
> I am not sure. The weight decay in particular looks like it won't be applied properly?

Sorry, to disturb your break!


I see your point -- weight decay definitely ought to be adjusted too, since it pushes the gradient further from the origin, in the direction of the size of the weights.
If we are maximising, that means we would be steping in the direction of the gradient, meaning we ought to to use `-\lambda * \theta _{t-1}` instead.

But for `m_t`, we have an weighted average of past gradients, so flipping the LR is equivalent to flipping the gradient. `v_t` is always positive, so we can ignore it.

I'd be happy to flip the gradient or to adjust the weight decay term. Which would you prefer? I'll update the docs accordingly and merge in the nightly changes

**Comment by albanD** [comment]
I feel like flipping the gradient is better as you have a change in 1 place. Otherwise you need to flip weight decay and lr: changes in 2 places.

File: torch/optim/_multi_tensor/adam.py [hunk 1]
```diff
@@ -79,6 +79,8 @@ def step(self, closure=None):
                 if p.grad is not None:
                     if p.grad.is_sparse:
                         raise RuntimeError('Adam does not support sparse gradients, please consider SparseAdam instead')
+                    if group['maximize']:
+                        p.grad = -p.grad
```
Comment (albanD) [inline torch/optim/_multi_tensor/adam.py:83]: @ngimel do we have a foreach neg that we would want to use here?

**Review by albanD — APPROVED** [review]
sounds good! Thanks for the update.

File: torch/optim/_multi_tensor/adam.py (hunk 1 above)
Comment (albanD) [reply torch/optim/_multi_tensor/adam.py:83]: Yes, nothing is differentiable here.

Comment (albanD) [reply torch/optim/_multi_tensor/adam.py:83]: @OliverFM could you use the foreach version to flip all of them at once?

Comment (OliverFM) [reply torch/optim/_multi_tensor/adam.py:83]: Not sure how this would work best.
Since `_foreach_neg` takes a list, or tuple, of tensor inputs, I'd need to generate this tuple, then apply the function, then loop through the parameters to add the gradients back.

Unless I am missing some smart way of doing a batch assignment of gradients to tensors.

Comment (albanD) [reply torch/optim/_multi_tensor/adam.py:83]: You should *not* modify the param's .grad field in this function. So there is no need to add the gradients back.
You can simply generate the gradient list as before and apply this function.

Comment (OliverFM) [reply torch/optim/_multi_tensor/adam.py]: made the changes!

**Review by albanD — APPROVED** [review]
Thanks for the update!
